### PR TITLE
[SIMD] dr_prediction_AVX optimization imrovement

### PR DESCRIPTION
**Pull request by Zvictoria**
PR #304  reincarnation. The third attempt to introduce AOM AV1 directional prediction AVX optimization changes to SVT AV1.
z2 functions performance is up to 2x better.

### CODE REVIEW

**Comment by 1480c1** [comment]
Wrong PR reference?

**Comment by 1480c1** [comment]
#304 instead?

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 1]
```diff
@@ -1151,7 +1154,7 @@ void intra_mode_vertical_luma_avx2_intrin(
     uint8_t            *ref_samples,             //input parameter, pointer to the reference samples
     uint8_t            *prediction_ptr,          //output parameter, pointer to the prediction
     const uint32_t      prediction_buffer_stride, //input parameter, denotes the stride for the prediction ptr
-    const EbBool     skip                    //skip one row
+    const EbBool     skip                    //skip one row
```
Comment (1480c1) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]:
```suggestion
    const EbBool     skip                    //skip one row
```
please run
`sed -i 's/[[:space:]]*$//' Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c`

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 2]
```diff
@@ -1263,19 +1268,21 @@ void intra_mode_dc_luma_avx2_intrin(
     uint8_t            *ref_samples,                 //input parameter, pointer to the reference samples
     uint8_t            *prediction_ptr,              //output parameter, pointer to the prediction
     const uint32_t      prediction_buffer_stride,     //input parameter, denotes the stride for the prediction ptr
-    const EbBool     skip)                       //skip one row
+    const EbBool     skip)                       //skip one row
```
Comment (1480c1) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: ditto

**Comment by Zvictoria** [comment]
@1480c1 thanks for correction! just a misprint from my side

**Comment by 1480c1** [comment]
Question @Zvictoria, what editor do you use when coding?

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 3]
```diff
@@ -9864,8 +10008,9 @@ static void highbd_dr_prediction_z3_16x16_avx2(uint16_t *dst, ptrdiff_t stride,
         dy);
     highbd_transpose16x16_avx2(dstvec, d);
 
-    for (int32_t i = 0; i < 16; i++)
+    for (int32_t i = 0; i < 16; i++) {
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: no braces needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 4]
```diff
@@ -9877,17 +10022,21 @@ static void highbd_dr_prediction_z3_32x32_avx2(uint16_t *dst, ptrdiff_t stride,
         dy);
 
     highbd_transpose16x16_avx2(dstvec, d);
-    for (int32_t j = 0; j < 16; j++)
+    for (int32_t j = 0; j < 16; j++) {
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: no braces needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 5]
```diff
@@ -9877,17 +10022,21 @@ static void highbd_dr_prediction_z3_32x32_avx2(uint16_t *dst, ptrdiff_t stride,
...
         _mm256_storeu_si256((__m256i *)(dst + j * stride), d[j]);
+    }
     highbd_transpose16x16_avx2(dstvec + 16, d);
-    for (int32_t j = 0; j < 16; j++)
+    for (int32_t j = 0; j < 16; j++) {
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: no braces needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 6]
```diff
@@ -9877,17 +10022,21 @@ static void highbd_dr_prediction_z3_32x32_avx2(uint16_t *dst, ptrdiff_t stride,
...
         _mm256_storeu_si256((__m256i *)(dst + j * stride + 16), d[j]);
+    }
     highbd_transpose16x16_avx2(dstvec + 32, d);
-    for (int32_t j = 0; j < 16; j++)
+    for (int32_t j = 0; j < 16; j++) {
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: no braces needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 7]
```diff
@@ -9877,17 +10022,21 @@ static void highbd_dr_prediction_z3_32x32_avx2(uint16_t *dst, ptrdiff_t stride,
...
         _mm256_storeu_si256((__m256i *)(dst + (j + 16) * stride), d[j]);
+    }
     highbd_transpose16x16_avx2(dstvec + 48, d);
-    for (int32_t j = 0; j < 16; j++)
+    for (int32_t j = 0; j < 16; j++) {
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: no braces needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 8]
```diff
@@ -9905,8 +10054,9 @@ static void highbd_dr_prediction_z3_16x32_avx2(uint16_t *dst, ptrdiff_t stride,
 
     highbd_dr_prediction_z1_32xN_internal_avx2(16, dstvec, left, upsample_left,
         dy);
-    for (int32_t i = 0; i < 32; i += 8)
+    for (int32_t i = 0; i < 32; i += 8) {
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: no braces needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 9]
```diff
@@ -9935,8 +10085,9 @@ static void highbd_dr_prediction_z3_32x16_avx2(uint16_t *dst, ptrdiff_t stride,
         dy);
     for (int32_t i = 0; i < 32; i += 16) {
         highbd_transpose16x16_avx2((dstvec + i), d);
-        for (int32_t j = 0; j < 16; j++)
+        for (int32_t j = 0; j < 16; j++) {
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: no braces needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 10]
```diff
@@ -9974,8 +10125,9 @@ static void highbd_dr_prediction_z3_64x16_avx2(uint16_t *dst, ptrdiff_t stride,
         dy);
     for (int32_t i = 0; i < 64; i += 16) {
         highbd_transpose16x16_avx2((dstvec + i), d);
-        for (int32_t j = 0; j < 16; j++)
+        for (int32_t j = 0; j < 16; j++) {
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: no braces needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 11]
```diff
@@ -10091,6 +10243,8 @@ void av1_highbd_dr_prediction_z3_avx2(uint16_t *dst, ptrdiff_t stride, int32_t b
     return;
 }
 
+
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: extra spaces not needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 12]
```diff
@@ -10420,8 +10574,10 @@ void aom_paeth_predictor_64x16_avx2(uint8_t *dst, ptrdiff_t stride,
   }
 }
 
+
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: extra spaces not needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 13]
```diff
@@ -10855,3 +11012,4 @@ void aom_highbd_paeth_predictor_2x2_avx2(uint16_t *dst, ptrdiff_t stride,
     *(uint32_t*)&dst[0] = _mm256_extract_epi32(row, 0);
     *(uint32_t*)&dst[stride] = _mm256_extract_epi32(row, 1);
 }
+
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: extra spaces not needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 14]
```diff
@@ -582,6 +584,7 @@ void intra_mode_planar_av1_avx2_intrin(
     }
 }
 
+
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: extra spaces not needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 15]
```diff
@@ -1481,10 +1491,13 @@ void intra_mode_dc_luma_avx2_intrin(
                 _mm_storel_epi64((__m128i *)(prediction_ptr + 2 * pStride), _mm_or_si128(_mm_and_si128(xmm_left, xmm_mask2), xmm_predictionDcValue));
 
                 _mm_storel_epi64((__m128i *)(prediction_ptr + 3 * pStride), _mm_or_si128(_mm_and_si128(_mm_srli_si128(xmm_left, 1), xmm_mask2), xmm_predictionDcValue));
+
             }
         }
     }
 
+
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: extra spaces not needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 16]
```diff
@@ -1511,20 +1524,24 @@ void intra_mode_dc_luma_avx2_intrin(
 
         xmm_sum256 = _mm256_castsi128_si256(xmm_sum1);
 
-        //#endif
+
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: extra spaces not needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 17]
```diff
@@ -1511,20 +1524,24 @@ void intra_mode_dc_luma_avx2_intrin(
...
+        //#endif      
         xmm_set = _mm256_castsi128_si256(_mm_set1_epi32(32));
 
         xmm_sum128_2 = _mm256_add_epi32(xmm_sum256, xmm_set); // add offset
         xmm_predictionDcValue = _mm256_srli_epi32(xmm_sum128_2, 6); //_mm256_srli_epi32
 
+
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: extra spaces not needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 18]
```diff
@@ -1511,20 +1524,24 @@ void intra_mode_dc_luma_avx2_intrin(
...
         __m128i dc128 = _mm256_castsi256_si128(xmm_predictionDcValue);
 
         uint8_t dc = _mm_cvtsi128_si32(dc128);
         xmm_predictionDcValue = _mm256_set1_epi8(dc);//_mm_broadcastb_epi8
 
+
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: extra spaces not needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 19]
```diff
@@ -1511,20 +1524,24 @@ void intra_mode_dc_luma_avx2_intrin(
...
         uint32_t count;
 
         for (count = 0; count < 2; ++count) {
+
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: extra spaces not needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 20]
```diff
@@ -1552,8 +1569,14 @@ void intra_mode_dc_luma_avx2_intrin(
             _mm256_storeu_si256((__m256i *)(prediction_ptr + 3 * pStride), xmm_predictionDcValue);
 
             prediction_ptr += (pStride << 2);
+
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: extra spaces not needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 21]
```diff
@@ -1552,8 +1569,14 @@ void intra_mode_dc_luma_avx2_intrin(
...
+
         }
+
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: extra spaces not needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 22]
```diff
@@ -1552,8 +1569,14 @@ void intra_mode_dc_luma_avx2_intrin(
...
+
     }
+
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: extra spaces not needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 23]
```diff
@@ -1900,6 +1926,8 @@ void aom_dc_top_predictor_64x16_avx2(uint8_t *dst, ptrdiff_t stride,
     row_store_64xh(&row, 16, dst, stride);
 }
 
+
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: extra spaces not needed

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 24]
```diff
@@ -2081,12 +2114,13 @@ void intra_mode_angular_2_avx2_intrin(
     uint8_t            *ref_samples,                 //input parameter, pointer to the reference samples
     uint8_t            *prediction_ptr,              //output parameter, pointer to the prediction
     const uint32_t      prediction_buffer_stride,     //input parameter, denotes the stride for the prediction ptr
-    const EbBool     skip)                       //skip one row
+    const EbBool     skip)                       //skip one row
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: trailing space

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 25]
```diff
@@ -2188,15 +2224,18 @@ void intra_mode_angular_18_avx2_intrin(
     uint8_t            *ref_samples,                 //input parameter, pointer to the reference samples
     uint8_t            *prediction_ptr,              //output parameter, pointer to the prediction
     const uint32_t      prediction_buffer_stride,     //input parameter, denotes the stride for the prediction ptr
-    const EbBool     skip)                    //skip one row
+    const EbBool     skip)                    //skip one row
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: trailing space

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 26]
```diff
@@ -2296,15 +2336,18 @@ void intra_mode_angular_34_avx2_intrin(
     uint8_t            *ref_samples,                 //input parameter, pointer to the reference samples
     uint8_t            *prediction_ptr,              //output parameter, pointer to the prediction
     const uint32_t      prediction_buffer_stride,     //input parameter, denotes the stride for the prediction ptr
-    const EbBool     skip)                       //skip one row
+    const EbBool     skip)                       //skip one row
```
Comment (hassount) [inline Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: trailing space

**Comment by hassount** [comment]
@Zvictoria thank you very much for cleaning up the patch making it a little easier to review. 
there are many extra braces / spaces / including trailing spaces that have been removed in previous commits that are being added back here. Could you please address those ?

File: Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c [hunk 27]
```diff
@@ -9319,49 +9271,50 @@ static void highbd_dr_prediction_z2_Nx8_avx2(
         int32_t y = r + 1;
         int32_t base_x = (-y * dx) >> frac_bits_x;
         int32_t base_shift = 0;
-        if (base_x < (min_base_x - 1))
+        if (base_x < (min_base_x - 1)) {
```
[the remark this answers is not in the document]
Comment (Zvictoria) [reply Source/Lib/Common/ASM_AVX2/EbIntraPrediction_Intrinsic_AVX2.c]: @1480c1 I use Visual Studio 2017. If there are some code editor settings for it to make my code compliant with SVT AV1 standard, could you please share?

**Comment by EwoutH** [comment]
Part of #262.

@hassount Could you add a _SIMD_ or _Assembly_ label to this PR in GitHub?

**Comment by Zvictoria** [comment]
@ttrigui
> There was no speed gain recorded when testing this code change for different encoder modes ( M6, M7 and M8) at three different rates ( qp values 32, 43 and 55).

We need to check if the functions modified (z2 directional prediction) were used in the encoding indeed). Also what frame sizes checked? It would be cool if you share your test setup (command line) with me.

**Comment by ttrigui** [comment]
@Zvictoria please fin below the command used.
sudo ./SvtAv1EncApp  -i  speed_1080p.yuv -w 1920 -h 1080 -fps 60 -b M8_speed.ivf -enc-mode 8 -scm 0 -q 55 1>speed_1080p_enc-mode8_q55.txt

**Comment by mafonsoN** [comment]
@Zvictoria @ttrigui any follow ups on this ?

**Comment by Zvictoria** [comment]
@mafonsoN , @ttrigui  I've done some checks myself finally. Unfortunately or fortunately for the particular encoding (I don't have speed_1080p file sorry, and would be happy to get any clue on where to get it) of some yuv 1080p file with w 1920 -h 1080 -fps 60 -enc-mode 8 -scm 0 -q 55 command line according to Intel Vtune Amplifier performance profiler the duration of dr_prediction functions (optimized in this patch ) is slightly more than 0%. Therefore  my optimization input is not seen. Some other test is necessary. However I do insist on this patch integration. First in makes code cleaner, shorter and more correct (data alignment added for all data, so some seldom but possible crashes for alignment read have gone). Second we do need SVT code to be consistent with AOM version for the common sense reason :) And just a friendly reminder - this code of mine is a part of AOM for half a year already.  Please tell me what else do you need to make this patch accepted.

**Comment by mafonsoN** [comment]
@Zvictoria Thansks for the tests. However, first, can you please rebase your branch to get the latest changes from master and fix the conflicts in EbIntraPrediction_Intrinsic_AVX2.c? Also, if you could please address the style issues?

**Comment by joelsole** [comment]
#528 closes this one.

**Pull request closed without merging**